add merchant module & map endpoints for adding new map entries.

islandbitcoin opened the pull request.

brh28 commented on src/app/merchants/approve-merchant-map.ts:
```diff
@@ -0,0 +1,13 @@
+): Promise<BusinessMapMarker | ApplicationError> => {
+  const merchantsRepo = MerchantsRepository()
+
+  const merchant = await merchantsRepo.findById(id)
```
I believe `findOneAndUpdate` does the full operation you're looking for in a single operation: https://mongoosejs.com/docs/tutorials/findoneandupdate.html

brh28 commented on src/app/merchants/delete-merchant-map.ts:
```diff
@@ -0,0 +1,36 @@
+  const merchants = await merchantsRepo.findByUsername(usernameChecked)
+  if (merchants instanceof Error) return merchants
+
+  if (merchants.length === 0) return true
```
this line appears to be unneeded

brh28 commented on src/app/merchants/index.ts:
It would be helpful to have types on the returns of these functions, especially since these are high level functions in the directory

brh28 commented on src/services/mongoose/accounts.ts:
```diff
@@ -76,34 +76,34 @@ export const AccountsRepository = (): IAccountsRepository => {
-        },
-        { username: 1, title: 1, coordinates: 1 },
-      )
+  // // FIXME: could be in a different file? does not return an Account
```
Let's delete the old code

brh28 commented on src/services/mongoose/merchants.ts:
```diff
@@ -0,0 +1,166 @@
+    try {
+      const result = await Merchant.findOne({ id })
+      if (!result) {
+        return new CouldNotFindMerchantFromIdError(id)
```
It's strange to me that empty queries are considered errors, but if that's how the rest of the codebase is handling it, I suppose we can follow the pattern

brh28 commented on src/services/mongoose/merchants.ts:
```diff
@@ -0,0 +1,166 @@
+    validated: boolean
+  }): Promise<BusinessMapMarker | RepositoryError> => {
+    try {
+      const location = {
```
Let's define this as our type.

islandbitcoin replied:
you mean Promise<BusinessMapMarker | RepositoryError> ?

brh28 commented on src/services/mongoose/merchants.ts:
```diff
@@ -0,0 +1,166 @@
+  }): Promise<BusinessMapMarker | RepositoryError> => {
+    try {
+      const location = {
+        type: "Point",
```
This should ideally reference an enum with the valid types

brh28 commented on src/services/mongoose/schema.ts:
```diff
@@ -296,6 +296,96 @@ AccountSchema.index({
+    unique: true,
+    sparse: true,
+    required: true,
+    default: () => crypto.randomUUID(),
```
is there a reason we're creating an id field rather than using Mongo's native `_id: ObjectID`?

islandbitcoin replied:
legacy code from Galoy

brh28 commented:
@islandbitcoin TBH, I can't give a good code review here without more context. 

Ideally, for changes of this scope, we should have the requirements (i.e the problem) clearly defined along with design diagrams that gives a blueprint to our approach (i.e solution). The code review is then just to verify the implementation (and testing) is adequate.    

Without a better conceptual understanding, I can only provide trivial feedback. If you're confident in the changes provided, I can approve it. For a more thorough review, I'd want to know more about the data and how it's Created, Retreived, Updated, Deleted, etc

islandbitcoin commented:
> @islandbitcoin TBH, I can't give a good code review here without more context.
> 
> Ideally, for changes of this scope, we should have the requirements (i.e the problem) clearly defined along with design diagrams that gives a blueprint to our approach (i.e solution). The code review is then just to verify the implementation (and testing) is adequate.
> 
> Without a better conceptual understanding, I can only provide trivial feedback. If you're confident in the changes provided, I can approve it. For a more thorough review, I'd want to know more about the data and how it's Created, Retreived, Updated, Deleted, etc

good call on all points. It's about time we move from hackathon to proper requirements gathering and implementation methods for all feature changes going forward. 

I'm mainly confident in this code change because it's an existing use case being used by Galoy today, and most of this is from their codebase.
